File: 09-backtracking/00079-Word_Search/00079-word_search.java
// 79: Word Search
// https://leetcode.com/problems/word-search


class Solution {
    public boolean backtrack(char[][] board, int i, int j, String word) {
        if (word.length() == 0) return true;
        if (i<0 || i>=board.length || j<0 || j>=board[0].length || board[i][j]!=word.charAt(0)) return false;
        
        char c = board[i][j];
        board[i][j] = '*';
        String s = word.substring(1);
        
        boolean result =  backtrack(board, i-1, j, s)
        || backtrack(board, i+1, j, s)
        || backtrack(board, i, j-1, s)
        || backtrack(board, i, j+1, s);
        board[i][j] = c;
        
        return result;
    }


    // SOLUTION
    public boolean exist(char[][] board, String word) {
        for (int i=0; i<board.length; i++)
            for (int j=0; j<board[0].length; j++)
                if (backtrack(board, i, j, word))
                    return true;
        
        return false;
    }


    public static void main(String[] args) {
        Solution s = new Solution();

        // INPUT :
        char[][] board = {
            {'A','B','C','E'},
            {'S','F','C','S'},
            {'A','D','E','E'}
        };
        String word = "ABCCED";

        // OUTPUT :
        System.out.println(s.exist(board, word));
    }
}

File: 09-backtracking/00079-Word_Search/00079-word_search.py
# 79: Word Search
# https://leetcode.com/problems/word-search


from typing import List

class Solution:
    # SOLUTION
    def exist(self, board: List[List[str]], word: str) -> bool:
        def backtrack(board: List[List[str]], i: int, j: int, word: str):
            if len(word) == 0 : return True
            if i<0 or i>=len(board) or j<0 or j>= len(board[0]) or board[i][j]!=word[0] : return False

            c = board[i][j]
            board[i][j] = '*'
            s = word[1:]

            result =   backtrack(board, i-1, j, s) \
                    or backtrack(board, i+1, j, s) \
                    or backtrack(board, i, j-1, s) \
                    or backtrack(board, i, j+1, s)
            board[i][j] = c

            return result


        for i in range(len(board)):
            for j in range(len(board[0])):
                if backtrack(board, i, j, word):
                    return True
        
        return False



if __name__ == "__main__":
    o = Solution()

    # INPUT
    board = [
        ["A","B","C","E"],
        ["S","F","C","S"],
        ["A","D","E","E"]
    ]
    word = "ABCCED"

    # OUTPUT
    result = o.exist(board, word)

    print(result)
